### PR TITLE
Add check for ignore overlaps at the start and end of a region

### PR DESCRIPTION
**Pull request by drjsanger**
@keiranmraine please check when you get the chance

### CODE REVIEW

**Comment by drjsanger** [comment]
@keiranmraine Additional changes added. Please check. Then we can ensure results match in CI

File: src/ignore_reg_access.c [hunk 1]
```diff
@@ -66,6 +66,21 @@ int ignore_reg_access_get_ign_reg_count_for_chr(char *ign_file, char *chr){
 	return -1;
 }
 
+seq_region_t *ignore_reg_access_get_ign_reg_inside(int pos, struct seq_region_t **regions, int entry_count){
+	int i=0;
+	for(i=0; i<entry_count; i++){
+		if(regions[i]->beg < pos && regions[i]->end > pos){
```
Comment (keiranmraine) [inline src/ignore_reg_access.c]: Doesn't this need to be:

```c
if(regions[i]->beg <= pos && regions[i]->end >= pos){
```

Comment (keiranmraine) [reply src/ignore_reg_access.c]: Otherwise the following values will return `NULL`

```c
beg = 10;
end = 20;
pos = 10;
```

Comment (drjsanger) [reply src/ignore_reg_access.c]: There are several different check methods. See
`ignore_reg_access_get_ign_reg_contained `

Comment (keiranmraine) [reply src/ignore_reg_access.c]: I'm having difficulty following this, I kind of feel like you only need 1 function to get the relevant ignore sections and a much simplified one to find the regions to actually retain... I could have this completely wrong though

Get all sections that overlap or are contained in one function:

```c
List *ignore_reg_access_get_ign_regs(int from, int to, struct seq_region_t **regions, int entry_count){
	List *li = List_create();
	int i=0;
	for(i=0; i<entry_count; i++){
	  if(from >= regions[i]->beg && from <= regions[i]->end
	    ||
	    end >= regions[i]->beg && end <= regions[i]->end)
	  ) {
			//Make a copy of this region and put in the list
			seq_region_t *reg_copy = malloc(sizeof(struct seq_region_t));
			check_mem(reg_copy);
			reg_copy->beg = regions[i]->beg;
			reg_copy->end = regions[i]->end;
			List_push(li,reg_copy);
		}
	}
	return li;
error:
	return NULL;
}
```

Now the all you need to do is loop over the retrieved list with a sepecial case on the first and last entries.

```c
List *ignore_reg_access_resolve_ignores_to_analysis_sections(int start, int end, struct seq_region_t **regions, int entry_count){
	List *li = ignore_reg_access_get_ign_regs(start,end,regions,entry_count);
  check(li != NULL,"Error fetching contained ignore regions.");
  
  List *reg_for_analysis = List_create();
 	seq_region_t *range = malloc(sizeof(struct seq_region_t));
 	
 	// slightly sudo code for is_first
 	is_first = 1;
 	LIST_FOREACH(li, first, next, cur){
 	  if(is_first) {
            is_first = 0;
 	    if(start >= ((seq_region_t *) cur->value)->beg) {
 	      range->beg = ((seq_region_t *) cur->value)->end+1
 	    }else{
 	      range->beg = start
 	    }
 	  }
		range->end = ((seq_region_t *) cur->value)->beg - 1;
		List_push(reg_for_analysis,range);
		range = malloc(sizeof(struct seq_region_t));
		if(((seq_region_t *) cur->value)->end < end) {
  		range->beg = ((seq_region_t *) cur->value)->end + 1;
  	}
	}
	if(range != NULL) { // if null the end of the last ignore region is >=end
	  range->end = end
	  List_push(reg_for_analysis,range);
	}

	List_clear_destroy(li);
	return reg_for_analysis;
error:
	List_clear_destroy(li);
	return NULL;

}
```

Comment (keiranmraine) [reply src/ignore_reg_access.c]: Ah this will also need the ability to create a region if `ignore_reg_access_get_ign_regs` returns an empty list.

Comment (keiranmraine) [reply src/ignore_reg_access.c]: I think just preceeding the loop with the following is all that is needed as the range would be overwritten in the loop if needed:

```c
range->beg = start
range->end = end
```

**Comment by drjsanger** [comment]
@kathryn-beal @rulixxx Could one of you please review if you get the chance. CI results for the commit ID linked above.

**Comment by rulixxx** [comment]
Took a look at all your changes. As far as I can tell code looks fine.

**Pull request merged**